### ISSUE TITLE
lint: only assignment is allowed in if init clause

### ISSUE THREAD

**Issue by cristaloleg** [go-critic/go-critic#492]
Name: `initClause` 
Before:
```go
if sideEffect(); cond {
}
```
After:
```go
sideEffect()
if cond {
}
```

Checker might be also extended for switches and probably for's

**Comment by Quasilyte** [comment]
Maybe `initSimpleStmt`? (Too bad spec doesn't provide other term for that.)
We can lookup more Go docs to figure out more conventional name for "init" short statement.

From spec: if/switch statements may contain `SimpleStmt`, and we usually want that to be used as "init" clause. Although it's not the case as the language permits any kind of "simple statement" there:
```
IfStmt = "if" [ SimpleStmt ";" ] Expression Block [ "else" ( IfStmt | Block ) ] .
SimpleStmt = EmptyStmt | ExpressionStmt | SendStmt | IncDecStmt | Assignment | ShortVarDecl
```

**Comment by Quasilyte** [comment]
Also note that this checker could handle both `if` and `switch` statements.

**Comment by Quasilyte** [comment]
Hah!
`go/ast` calls that simple statement "init":
```go
type IfStmt struct {
        If   token.Pos // position of "if" keyword
        Init Stmt      // initialization statement; or nil
        Cond Expr      // condition
        Body *BlockStmt
        Else Stmt // else branch; or nil
}
```

**Comment by cristaloleg** [comment]
UPD name is `initClause`

**Comment by Quasilyte** [comment]
We invented a new term. :sweat_smile:
